### PR TITLE
[SPARK-8844][SPARKR] head/collect is broken in SparkR.

### PR DESCRIPTION
**Pull request by sun-rui**
This is a WIP patch for SPARK-8844  for collecting reviews.

This bug is about reading an empty DataFrame. in readCol(), 
      lapply(1:numRows, function(x) {
does not take into consideration the case where numRows = 0.

Will add unit test case.

### CODE REVIEW

File: R/pkg/R/deserialize.R [hunk 1]
```diff
@@ -174,11 +174,46 @@ readRow <- function(inputCon) {
...
+readCol <- function(inputCon, numRows, colType) {
+  if (numRows > 0) {
+    # sapply can not work with POSIXlt
+    do.call(c, lapply(1:numRows, function(x) {
+      value <- readObject(inputCon)
+      # Replace NULL with NA so we can coerce to vectors
+      if (is.null(value)) NA else value
+    }))
+  } else {
+    vector(mode = colType)
```
Comment (shivaram) [inline R/pkg/R/deserialize.R]: So fundamentally it looks like this we want to create a vector with `mode = colType` if the column is empty. But does it matter if we get the `colType` wrong ? i.e. what happens if we just use the default mode (which I think is `logical`)

**Comment by shivaram** [comment]
cc @davies

**Comment by shivaram** [comment]
@sun-rui Any update on this ?

**Comment by sun-rui** [comment]
@shivaram , as you pointed out, I come with a simpler fix. I realized that simply creating an empty vector using vector() without mode is OK when there is no row in columns. Because vector() creates a zero-length vector of type "logical", and "logical" is of virutally lowest precedence in the R's type hierarchy, so it is supposed that "logical" type won't affect later coercion in operations on a data.frame of 0 row, for example, rbind().

**Comment by shivaram** [comment]
Jenkins, retest this please

**Comment by shivaram** [comment]
Thanks @sun-rui -- this fix is simple and looks good. Couple of minor things
1. There is an empty file `nohup.out` added in the PR which needs to be removed
2. Could we also call `head` in the test case ?

**Comment by sun-rui** [comment]
@shivaram, fixed.

**Comment by davies** [comment]
LGTM

**Comment by shivaram** [comment]
LGTM. Merging this

**Pull request closed without merging**